components/mediation/data-publishers/org.wso2.micro.integrator.observability/src/main/java/org/wso2/micro/integrator/observability/handler/metrichandler/MetricHandler.java

```diff
@@ -0,0 +1,440 @@
+                loadDefaultPrometheusReporter();
+            }
+        } else {
+            loadDefaultPrometheusReporter();
+        }
+    }
+
+    /**
+     * Load the PrometheusReporter class by default.
+     */
+    private void loadDefaultPrometheusReporter() {
+        try {
+            loadedMetricClass = PrometheusReporter.class;
+            metricReporterInstance = (MetricReporter) loadedMetricClass.newInstance();
+            log.debug("The class org.wso2.micro.integrator.obsrvability.handler.metrics.publisher.prometheus." +
+                    "reporter.PrometheusReporter loaded successfully");
+        } catch (IllegalAccessException | InstantiationException e) {
+            handleException("Error in loading the PrometheusReporter Class", e);
+        }
+    }
+
+    @Override
+    public boolean handleRequestInFlow(MessageContext synCtx) {
+        org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) synCtx)
+                .getAxis2MessageContext();
+
+        int serviceInvokePort ;
+        int internalHttpApiPort;
+
+        if ((null != synCtx.getProperty(SynapseConstants.PROXY_SERVICE))) {
+            String proxyName = axis2MessageContext.getAxisService().getName();
+
+            incrementProxyCount(proxyName);
+            startTimers(synCtx, proxyName, SynapseConstants.PROXY_SERVICE_TYPE, null);
+        } else if (null != synCtx.getProperty(SynapseConstants.IS_INBOUND)) {
+            String inboundEndpointName = synCtx.getProperty(SynapseConstants.INBOUND_ENDPOINT_NAME).toString();
+
+            incrementInboundEndPointCount(inboundEndpointName);
+            startTimers(synCtx, inboundEndpointName, MetricConstants.INBOUND_ENDPOINT,
+                    null);
+        } else if ((null != axis2MessageContext.getProperty(MetricConstants.TRANSPORT_IN_URL) &&
+                !axis2MessageContext.getProperty(MetricConstants.TRANSPORT_IN_URL).toString().
+                        contains("services"))) {
+            try {
+                serviceInvokePort = getServiceInvokePort(synCtx);
+                internalHttpApiPort = getInternalHTTPInboundEndpointPort(synCtx);
+
+                if (serviceInvokePort != internalHttpApiPort) {
+                    String context = axis2MessageContext.getProperty(MetricConstants.TRANSPORT_IN_URL).
+                            toString();
+                    String apiInvocationUrl = axis2MessageContext.getProperty(MetricConstants.SERVICE_PREFIX).
+                            toString() + context.replaceFirst(DELIMITER, EMPTY);
+                    String apiName = getApiName(context, synCtx);
+                    if (apiName != null) {
+                        incrementAPICount(apiName, apiInvocationUrl);
+                        startTimers(synCtx, apiName, SynapseConstants.FAIL_SAFE_MODE_API, apiInvocationUrl);
+                    }
+                }
+            } catch (NullPointerException ex) {
```

Comment:
Shall we handle NPE without catching it?

Reply:
Fixed